discopy/tensor.py

```diff
@@ -541,16 +540,17 @@ def __setstate__(self, state):
             state['data'] = state['_array']
             del state["_array"]
         super().__setstate__(state)
-        if self.dtype is None:
+        if self.dtype is None and self.data:
```

Comment:
Otherwise this will fail to infer `int` with `self.data = 0`!

```suggestion
        if self.dtype is None and self.data is not None:
```